### ISSUE TITLE
IsolateNameServer crashes on release mode with flutter 3.3.x but works with flutter 3.0.x

### ISSUE THREAD

**Issue by cool2apps** [flutter/flutter#113825]
## Steps to Reproduce

I am using ```IsolateNameServer``` to handle background messages comes from firebase cloud messaging. When app receives background message from FCM, ```_firebaseMessagingBackgroundHandler``` was executed properly until ```sendPort.send(message)```. At ```sendPort.send``` app crashes without giving any visual clue or any log. The same code works as expected with Flutter 3.0.5, but does not work with any version of flutter 3.3.x.

Similar problem was reported here 2 months ago and closed as fixed. But it looks like similar problem still exists: https://github.com/flutter/flutter/issues/109248

PS: It is not easy to publish whole source code. Because it is not an open source code project and it has about 20K lines of dart code. It includes a lot of project specific codes like firebase configuration and cloud code scripts to send FCM messages.

**Expected results:** 
 Program execution should continue with ```Shared.debug("new firebase cloud message from port")```

**Actual results:** 
 App crashes without any visual outcome or any log.

<details>
<summary>Code sample</summary>

``` 
  @pragma('vm:entry-point')
  static Future<void> _firebaseMessagingBackgroundHandler(RemoteMessage message) async {
    try {
      Shared.firebaseApp ??= await Firebase.initializeApp(
        options: DefaultFirebaseOptions.currentPlatform,
      );
      Shared.debug("Handling a background message");
      Shared.debug("background message sending to port");
      final SendPort? sendPort = IsolateNameServer.lookupPortByName('firebaseCloudMessagePort');
      if (sendPort == null) {
        Shared.debug("sendPort is null. message not send");
      } else {
        Shared.debug("sending message to sendPort");
        sendPort.send(message);
      }
    } catch (error) {
      Shared.debug("Error on _firebaseMessagingBackgroundHandler: "+error.toString());
    }
  }
...
    ReceivePort receivePort = ReceivePort();
    IsolateNameServer.registerPortWithName(receivePort.sendPort, 'firebaseCloudMessagePort');
    receivePort.listen((dynamic data) async {
      Shared.debug("new firebase cloud message from port");
      await Shared.handleFirebaseCloudMessage(data, "terminated", updateView);
      if (Shared.notificationView != null) {
        setView(Shared.notificationView!);
        Shared.notificationView = null;
      }
    });
```

</details>

<details>
  <summary>Logs</summary>

```
flutter run --release
Installing build/app/outputs/flutter-apk/app.apk...                34.4s
Flutter run key commands.
h List all available interactive commands.
c Clear the screen
q Quit (terminate the application on the device).
D/FlutterLocationService( 3869): Creating service.
D/FlutterLocationService( 3869): Binding to location service.
I/flutter ( 3869): *** RELEASE *** Notifications.init started
W/FlutterJNI( 3869): FlutterJNI.loadLibrary called more than once
I/flutter ( 3869): *** RELEASE *** No initial message
W/FlutterJNI( 3869): FlutterJNI.prefetchDefaultFontManager called more than once
W/FlutterJNI( 3869): FlutterJNI.init called more than once
I/flutter ( 3869): *** RELEASE *** Handling a background message
I/flutter ( 3869): *** RELEASE *** background message sending to port
I/flutter ( 3869): *** RELEASE *** sending message to sendPort
```

```
flutter analyze
Analyzing armdate...                                                    
No issues found! (ran in 6.5s)
```

```
flutter doctor -v
[✓] Flutter (Channel stable, 3.3.5, on Ubuntu 22.04.1 LTS 5.15.0-50-generic, locale en_US.UTF-8)
    • Flutter version 3.3.5 on channel stable at /home/coolapps/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d9111f6402 (2 days ago), 2022-10-19 12:27:13 -0700
    • Engine revision 3ad69d7be3
    • Dart version 2.18.2
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /home/coolapps/development/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /home/coolapps/development
    • Java binary at: /home/coolapps/development/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2021.3)
    • Android Studio at /home/coolapps/development/android-studio
    • Flutter plugin version 70.2.3
    • Dart plugin version 213.7433
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] Connected device (3 available)
    • SM M526BR (mobile)           • 192.168.0.29:35611 • android-arm64 • Android 12 (API 31)
    • sdk gphone64 x86 64 (mobile) • emulator-5556      • android-x64   • Android 13 (API 33) (emulator)
    • Linux (desktop)              • linux              • linux-x64     • Ubuntu 22.04.1 LTS 5.15.0-50-generic

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

</details>

**Comment by cool2apps** [comment]
> @cool2apps Does the behavior persist without using `firebase_messaging` plugin ? I am trying to understand if the crash you are reporting occur without any other plugin and simply using `IsolateNameServer` class.
> 
> When the crash occur, is your app in background or in terminated state ? If in background, are you removing the registered port ? Then in background isolate check if it exists first and handle the method, something like:
> 
> ```
> `_foregroundReceivePort!.listen((message) {
> _fcmMessageHandlerForeground(message);
> }
> ```
> 
> > At `sendPort.send` app crashes without giving any visual clue or any log.
> 
> Can you try by using `firebase_crashlytics` inside `_firebaseMessagingBackgroundHandler` and see if you get any logs that'll be helpful to debug further ?

1. App crashes on both states
2. If it is in background state, app crashes after printing ```sending message to sendPort``` debug message. I assume it executes and crashes after executing ```sendPort.send(message)```. But I don't have proof it was executed or not.
3. If it is terminated state, app crashes after printing ```sendPort is null. message not send``` debug message.
4. I don't remove registered port
5. I don't know if this problem exists without firebase_messaging. Because I am trying to bring app to foreground with FCM.
6. I don't listen foreground port. I listen FCM in foreground like this
```
    FirebaseMessaging.onMessage.listen((RemoteMessage message) async {
      Shared.debug('Got a message whilst in the foreground!');
      Shared.debug('Message data: ${message.data}');
      Shared.handleFirebaseCloudMessage(message, "foreground", updateView);
    });
```
7. I will try to use ```firebase_crashlytics``` and report back as soon as I get results.

Thanks

**Comment by mraleph** [comment]
It's likely a duplicate of https://github.com/flutter/flutter/issues/113183. I suggest searching through `adb logcat` for native crash reports to see if there are any. 

If possible try running the same application on master.

**Comment by cool2apps** [comment]
I added ```@pragma('vm:entry-point')``` for each class and for each function involves to ```IsolateNameServer``` but app crashes again. I suspect that the function sends is a static class. Is that causes any problem?

**Comment by vmtb** [comment]
It is related to the type of data sent by the SentPort.

**Comment by cool2apps** [comment]
> It is related to the type of data sent by the SentPort.

Which data can ben sent? I need to send RemoteMessage type.

**Comment by vmtb** [comment]
It'll be complicated

![image](https://user-images.githubusercontent.com/35667457/197252420-8954fc3c-681d-48aa-9c25-cd25417e3190.png)

**Comment by vmtb** [comment]
It is here https://api.flutter.dev/flutter/dart-isolate/SendPort/send.html

**Comment by cool2apps** [comment]
> It is here https://api.flutter.dev/flutter/dart-isolate/SendPort/send.html
Is it changed from flutter 3.0.5 to flutter 3.3.x? Because it is working perfect with flutter 3.0.5.

**Comment by vmtb** [comment]
Yes it works on previous version of flutter but 3.3.x causes the problem

**Comment by cool2apps** [comment]
> Yes it works on previous version of flutter but 3.3.x causes the problem

I guess there should be a bug with 3.3.x.

**Comment by vmtb** [comment]
Did you use flutter 3.3.1 previously on your machine?

**Comment by vmtb** [comment]
Have you ever use this version on you machine? (3.3.1)

**Comment by cool2apps** [comment]
I don't remember. Probably. I generally upgrade every new stable release of flutter until I noticed this problem.

**Comment by cool2apps** [comment]
> Have you ever use this version on you machine? (3.3.1)

What is special about 3.3.1 version?

**Comment by vmtb** [comment]
yes if you downgrade your flutter to 3.3.1 et add package [flutter_isolate: 2.0.2](https://pub.dev/packages/flutter_isolate) to your pubspeck, it might work. 

> > Have you ever use this version on you machine? (3.3.1)
> 
> What is special about 3.3.1 version?

**Comment by cool2apps** [comment]
> yes if you downgrade your flutter to 3.3.1 et add package [flutter_isolate: 2.0.2](https://pub.dev/packages/flutter_isolate) to your pubspeck, it might work.
> 
> > > Have you ever use this version on you machine? (3.3.1)
> > 
> > 
> > What is special about 3.3.1 version?

I will try and share the result. Thanks

**Comment by cool2apps** [comment]
> yes if you downgrade your flutter to 3.3.1 et add package [flutter_isolate: 2.0.2](https://pub.dev/packages/flutter_isolate) to your pubspeck, it might work.
> 
> > > Have you ever use this version on you machine? (3.3.1)
> > 
> > 
> > What is special about 3.3.1 version?

flutter_isolate 2.0.3 Fix release mode crashes on Flutter >3.3.0 - @pragma('vm:entry-point') must be added as a decorator to all entrypoint functions

v2.0.3 might be better?

**Comment by cool2apps** [comment]
> yes if you downgrade your flutter to 3.3.1 et add package [flutter_isolate: 2.0.2](https://pub.dev/packages/flutter_isolate) to your pubspeck, it might work.
> 
> > > Have you ever use this version on you machine? (3.3.1)
> > 
> > 
> > What is special about 3.3.1 version?

Tried but didn't work.

**Comment by cool2apps** [comment]
If I convert message to String, I can send it with sendPort without crashing the app, but I need to convert it back to RemoteMessage type at the other end. But I can't covert it to back to RemoteMessage. I guess my json knowledge is not well enough.
```
        var tmpData = message.data["data"];
        Map<String, dynamic> data = jsonDecode(tmpData);
        var stringMessage = data.toString();
        sendPort.send(stringMessage);
```
When I try to covert it with ```var tmp = jsonEncode(message) as Map<String, dynamic>;``` I get ```type '_OneByteString' is not a subtype of type 'Map<String, dynamic>' in type cast``` exception error.

**Comment by cool2apps** [comment]
I have just found a workaround. I am not sure if it is fully solved the problem. I will test more later.

At the sender part, I send ```message.data``` instead of ```message``` with.
```
sendPort.send(message.data);
```
At the receiver part, I generate message like this:
```
        RemoteMessage remoteMessage = RemoteMessage(data: data);
```
It looks like working for now for android. I need to test it with iOS too.

**Comment by aam** [comment]
@cool2apps how do you spawn your worker isolate(sender)?
Note that you are limited in kind of data you can send between isolates spawned using `Isolate.spawn` vs `Isolate.spawnUri`.
From https://api.flutter.dev/flutter/dart-isolate/SendPort/send.html:
```If the sender and receiver isolate share the same code (e.g. isolates created via [Isolate.spawn](https://api.flutter.dev/flutter/dart-isolate/Isolate/spawn.html)), the transitive object graph of message can contain any object, with the following exceptions:```

Note that fix for https://github.com/flutter/flutter/issues/109248 https://dart-review.googlesource.com/c/sdk/+/255261 was to ensure that above-quoted check is not accidentally bypassed by IsolateNameServer.

**Comment by aam** [comment]
> I don't spawn. Do I have to? As mentioned in sample code in original post, there are 2 codes running on same apk.

I see, thanks. I think FirebaseMessaging background message handler does end up with similar messaging limitations as what is outlined for `Isolate.spawnUri`, so you are limited to objects outlined in ```The transitive object graph of message can contain the following objects:``` paragraph from https://api.flutter.dev/flutter/dart-isolate/SendPort/send.html. This is similar to DownloadWorker mentioned in https://github.com/flutter/flutter/issues/109248#issuecomment-1215772666

**Comment by cool2apps** [comment]
> > I don't spawn. Do I have to? As mentioned in sample code in original post, there are 2 codes running on same apk.
> 
> I see, thanks. I think FirebaseMessaging background message handler does end up with similar messaging limitations as what is outlined for `Isolate.spawnUri`, so you are limited to objects outlined in `The transitive object graph of message can contain the following objects:` paragraph from https://api.flutter.dev/flutter/dart-isolate/SendPort/send.html. This is similar to DownloadWorker mentioned in [#109248 (comment)](https://github.com/flutter/flutter/issues/109248#issuecomment-1215772666)

But there was no such limitations with flutter 3.0.5. Is that new with flutter 3.3.x?

**Comment by dnfield** [comment]
I think the piece we shoudl fix in this case is making Android's `new FlutterEngine` by default spawn an engine in the same isolate group. But this would probably be breaking.

**Comment by dnfield** [comment]
Ahh interesting. @cool2apps I think you can work around this in your own app. Can you try this? Note that this is pretty hacky but I think it should work:

```java
// In MyApplication.java

  @Override
  public void onCreate() {
    FlutterJNI.Factory newThenSpawnFactory = new FlutterJNI.Factory() {
      FlutterJNI flutterJni = new FlutterJNI();
      boolean firstCall = false;
      @Override
      public FlutterJNI provideFlutterJNI() {
        if (firstCall) {
          firstCall = false;
          return flutterJni;
        }
        return flutterJni.spawn(null, null, null, null);
      }
    };
    FlutterInjector.setInstance(
          new FlutterInjector.Builder()
              .setFlutterJNIFactory(newThenSpawnFactory)
              .build());
    // ...
  }
```

You'll then need to update your AndroidManifest to use that if you're not already overriding Application somewhere, and you'll have to translate it ot kotlin if that's what you're using. Anyway, I think something like that will fix your issue.

**Comment by cool2apps** [comment]
> Ahh interesting. @cool2apps I think you can work around this in your own app. Can you try this? Note that this is pretty hacky but I think it should work:
> 
> ```java
> // In MyApplication.java
> 
>   @Override
>   public void onCreate() {
>     FlutterJNI.Factory newThenSpawnFactory = new FlutterJNI.Factory() {
>       FlutterJNI flutterJni = new FlutterJNI();
>       boolean firstCall = false;
>       @Override
>       public FlutterJNI provideFlutterJNI() {
>         if (firstCall) {
>           firstCall = false;
>           return flutterJni;
>         }
>         return flutterJni.spawn(null, null, null, null);
>       }
>     };
>     FlutterInjector.setInstance(
>           new FlutterInjector.Builder()
>               .setFlutterJNIFactory(newThenSpawnFactory)
>               .build());
>     // ...
>   }
> ```
> 
> You'll then need to update your AndroidManifest to use that if you're not already overriding Application somewhere, and you'll have to translate it ot kotlin if that's what you're using. Anyway, I think something like that will fix your issue.
> 
> This should cause invocations of `new FlutterEngine` to always use a spawned engine, which will share heaps and allow for sendport usage like you want.

Thanks but I don't want to hack if it is not mandatory. Sending ```message.data``` instead of ```message``` looks like working right now.

**Comment by a-siva** [comment]
> I think the piece we shoudl fix in this case is making Android's `new FlutterEngine` by default spawn an engine in the same isolate group. But this would probably be breaking.

Based on this comment it appears this is an engine issue so dropping the dart-dependency label.

**Comment by iazel** [comment]
Perhaps this issue is related to #147469 ?

If the isolate exists before calling `IsolateNameServer.removePortNameMapping`, then you'll see this issue the next time you retrieve that port. This is often the case during hot-restarts.

It is easy to verify if you are having this issue, check the result of `IsolateNameServer.registerPortWithName`, if it is `true`, then you have it. Keep in mind that registering a port **will not** overwrite it in case the name is already used.

Another reason, although quite unlikely to occur, is that, according to code sample, port is first registered and then listened. If you have a lot of users, but this error happens seldomly, then that could be the cause.
